client/asset/dcr/dcr.go

```diff
@@ -2208,36 +2287,65 @@ func (dcr *ExchangeWallet) ValidateSecret(secret, secretHash []byte) bool {
 	return bytes.Equal(h[:], secretHash)
 }
 
-func (dcr *ExchangeWallet) coinConfirmations(ctx context.Context, id dex.Bytes) (confs uint32, spent bool, err error) {
-	txHash, vout, err := decodeCoinID(id)
+// SwapConfirmations gets the number of confirmations and the spend status for
+// the specified swap. The contract and matchTime are provided so that wallets
+// may search for the coin using light filters.
+//
+// If the swap was not funded by this wallet, and it is already spent, this
+// method may return asset.CoinNotFoundError. Compare dcr.externalTxOut.
+//
+// If the coin is located, but recognized as spent, no error is returned.
+func (dcr *ExchangeWallet) SwapConfirmations(ctx context.Context, coinID, contract dex.Bytes, matchTime time.Time) (confs uint32, spent bool, err error) {
+	txHash, vout, err := decodeCoinID(coinID)
 	if err != nil {
 		return 0, false, err
 	}
-	// Check for an unspent output.
-	txOut, _, err := dcr.getTxOut(txHash, vout, true)
-	if err == nil && txOut != nil {
-		return uint32(txOut.Confirmations), false, nil
+
+	// Check if we can find the contract onchain without using cfilters.
+	_, confs, _, spent, err = dcr.lookupTxOutput(ctx, txHash, vout)
+	if err == nil {
+		return confs, spent, nil
+	} else if err != asset.CoinNotFoundError {
+		return 0, false, err
 	}
-	// Check wallet transactions.
-	tx, err := dcr.wallet.GetTransaction(ctx, txHash)
+
+	// Prepare the pkScript to find the contract output using block filters.
+	scriptAddr, err := stdaddr.NewAddressScriptHashV0(contract, dcr.chainParams)
+	if err != nil {
+		return 0, false, fmt.Errorf("error encoding script address: %w", err)
+	}
+	_, p2shScript := scriptAddr.PaymentScript()
+
+	// Find the contract and it's spend status using block filters.
+	dcr.log.Debugf("Contract output %s:%d NOT yet found, will attempt finding it with block filters.", txHash, vout)
+	confs, spent, err = dcr.lookupTxOutWithBlockFilters(ctx, newOutPoint(txHash, vout), p2shScript, matchTime)
 	if err != nil {
 		return 0, false, err
 	}
-	return uint32(tx.Confirmations), true, nil
+	return confs, spent, err
 }
 
-// SwapConfirmations gets the number of confirmations for the specified coin ID
-// by first checking for a unspent output, and if not found, searching indexed
-// wallet transactions.
-func (dcr *ExchangeWallet) SwapConfirmations(ctx context.Context, coinID dex.Bytes, _ dex.Bytes, _ time.Time) (confs uint32, spent bool, err error) {
-	return dcr.coinConfirmations(ctx, coinID)
+func (dcr *ExchangeWallet) confirms(blockHeight int64) uint32 {
```

Comment:
`confirms` is only used by `lookupTxOutWithBlockFilters`, which is also in a different file (externaltx.go).  It's also a bit specialized in that it converts to uint32 and falls back to the cached best block.

So that suggests to me this might be better off in `lookupTxOutWithBlockFilters` where it can prevent a negative `int64` being cast into a huge `uint32`.  The ctx arg can be used instead that way too.